Introduce event metadata and UTC date formatting

atenart opened the pull request:
This adds event metadata to the events file allowing non-event information to be stored for post-processing. As an example, the Retis version is included as metadata. I decided not to follow the proposal in #179 to allow lazy parsing of events. This looks like:

```
{"metadata":{"common":{"retis_version":"v1.4.0-42c293cc65eb"}}}
{"common":{"smp_id":9,"task":{"comm":"swapper/9","pid":0,"tgid":0},"timestamp":10291966331379}, ...}
...
```

At core metadata events are a special case of events but reuse the same logic and most of the events implementation. I've decided to still split the two types but we could just make them a normal event and reuse even more code (minor modifications to the Event implementation would be needed).

The file factory can now return normal and metadata events as part of its processing. I've decided not to force metadata events to be special (eg. only one in the event file or being on top of the normal events). This could allow metadata to be filled in flight or to be overridden later on.

PR is not fully ready (tests weren't converted) but is enough to start the discussion.

Fixes #179, #299.

amorenoz commented on src/core/events/metadata.rs:
```diff
@@ -0,0 +1,54 @@
+///
+/// We reuse here the Event object internally as we're storing the exact same
+/// kind of data (map of ModuleId to custom structs) with the same set of
+/// capabilities (JSON conversion, marshalling, etc). However we redefine our
```
Doesn't this approach limit the scope of data that we can store here?
Most of the uses we've found to the metadata section are "common", where as it's unlikely other modules will need to store metadata associated with them (maybe you've thought of a use-case I fail to see?).

Therefore, this is a limitation originated by internal implementation details (e.g: trying to reuse code) that ends up causing most of the information to unnecessarily be inside a "common".

amorenoz commented on src/core/events/file.rs:
```diff
@@ -1,42 +1,94 @@
-impl FileEventsFactory {
-    pub(crate) fn new<P>(file: P) -> Result<Self>
+impl FileEventsFactory<'_> {
+    pub(crate) fn new<P>(file: P, metadata_factories: MetadataFactories) -> Result<Self>
```
What's the use case for allowing the modules to insert module-specific metadata all scattered throughout the event file? What kind of data cannot be stored in an actual event or in a global (single) metadata section?

amorenoz replied:
Also, please comment on #388 as it goes in the opposite direction regarding Factories.

amorenoz commented on src/core/events/bpf.rs:
```diff
@@ -430,6 +430,16 @@ impl EventFmt for CommonMetadata {
+
+impl RawEventSectionFactory for CommonMetadataFactory {
+    fn from_raw(&mut self, _: Vec<BpfRawSection>) -> Result<Box<dyn EventSection>> {
+        bail!("Can't generate metadata information from raw data");
```
This is, IMHO, another example of why the current Factory API  does not work.

atenart commented:
I'm not going to answer comments individually, but we discussed this offline and here is the takeaway:

- Metadata events are just normal events, we'll not differentiate them.
- This should look way cleaner on top of the event rework from #388.

Next step will be to rebase and rework this on top of the event rework, once ready.

atenart commented:
Based on #388. New version just reuses the current `Event` and its sections to add metadata. Metadata sections are just sections and behave the same. The only difference for now is those sections are left out of the current print formats (not displayed by `collect`, `print` and `sort`.

A few other improvements in this PR, linked to how to display events.

atenart commented:
c8s runtime test failed for an unrelated reason. Same as #388, we can put it aside for now.

atenart commented:
On top of the event metadata additions, pushed the first user: UTC date formatting. This improves the PR I believe as it shows a first user of the feature.

While at it, made a few changes to the way we control the format when displaying events.

amorenoz commented:
Also, we might need a section in the docs for the new event section?

atenart commented on retis-events/src/events.rs:
```diff
@@ -122,13 +122,11 @@ impl Event {
+            write!(f, "{}", common.display(format))?;
+        }
         if let Some(kernel) = self.0.get(&SectionId::Kernel) {
             write!(f, " {}", kernel.display(format))?;
```
In practice that can't happen but because no event section is mandatory now the code need to take it into account. You're right, we need to handle that case and not have a leading space. This can be done with `DelimWriter`.

However there's another issue: if a metadata event is displayed, it won't have any of the first sections and will be intended which is not what we want (I forgot to fix that), eg,

```
  Retis version v1.4.0-768e78200421-dirty

132005966153185 (0) [swapper/0] 0 [tp] net:netif_rx ...
```

This might be possible to solve with a `DelimWriter` too, forcing the first sub-section to be printed to not be intended. But if that first section is `skb` that might be a little surprising. The root cause is event sections are quite generic, but not in the way we print them, we have to make some assumptions. I think that's fine though, printing some of the first sections in a nice way is more an optimization than something we are forced to do.

Note: for the above to work the stack section might need some special handling, or be moved to the end of the output.

amorenoz replied:
+1 to use `DelimWriter`

atenart replied:
With the common section staying mandatory, the above is no longer a concern.

atenart commented on retis-events/src/events.rs:
```diff
@@ -122,13 +122,11 @@ impl Event {
-        )?;
+        // First format the first event line starting with the {common} section,
+        // followed by the {kernel} or {user} one.
+        if let Some(common) = self.0.get(&SectionId::Common) {
```
On one side I do agree this is not common to all events, but it should be at least to most. The second thing is this is kinda part of an ABI (at least the "common" keywork in JSON), even though we could break it.

Another option: the common section has a timestamp and optional information (the smp id is not but setting 0 is perfectly fine). Maybe we actually do want it to be mandatory, so even for metadata events we have a timestamp information (eg. `Event::with_common()` to create an event for non-BPF events)?

amorenoz replied:
Hmm... zero is still a valid smp id so I don't know if we should do that. Maybe having it `Optional`?
Other than that, I think it's a good idea. I don't really like having a json object with just one field but I think it's better than having a "common" section that is not common.

Some other thought (trying to think of other pieces of information that might come handy in a "common" section): Do we need to flag an event as a metadata one? Events coming from eBPF side won't have metadata sections right? If a collector has to add data that is related to it, it can create another (normal) section.

So, maybe we need a `type=[bpf|meta]` field? or even a `meta=[true|false] # default false to avoid extra space in normal events which are more common`?

amorenoz replied:
That way we can easily identify an event as containing only metadata without having to iterate through sections.

atenart replied:
> Hmm... zero is still a valid smp id so I don't know if we should do that. Maybe having it `Optional`? Other than that, I think it's a good idea. I don't really like having a json object with just one field but I think it's better than having a "common" section that is not common.

Sounds good.

> Some other thought (trying to think of other pieces of information that might come handy in a "common" section): Do we need to flag an event as a metadata one? Events coming from eBPF side won't have metadata sections right? If a collector has to add data that is related to it, it can create another (normal) section.
> 
> So, maybe we need a `type=[bpf|meta]` field? or even a `meta=[true|false] # default false to avoid extra space in normal events which are more common`?

(Note that before writing this down I already made the common section to be mandatory only).

That would shift the metadata vs non-metadata to the event instead of the section level. My idea was really not to differentiate "metadata" vs other sections. However I do agree we *do* make a distinction as of now (more on that later).

I don't think an event should be metadata only or metadata free. The metadata is there to provide additional information for post-processing. It can be global information (that can be extended / amended by future events) but it could be specific to a given event.

Taking a step back, the main points to define to get an agreement here are: what is an event and what is a metadata section.

For the metadata part, my view is this should be at the section level. The current distinction based on the section id might not be the best fit, neither is the "metadata" word we use in some places. IMO the only difference is the ability for `EventFmt` to produce something. Maybe we should only make the distinction here (it's already done in `PrintSingle/Series` for example): a "metadata section" would be a normal section not producing any direct output.

For the event part, I think it has to bet loosely defined to get some flexibility. To me an event is a set of sections somehow related to each others. It might be to a packet event (aka. probe + timestamp) or to the initial configuration that is dumped by collectors (aka. timestamp only).

The difference in how those are consumed at post-processing time (is it global info? specific to an event?) is really only depending on the kind of section it is. That's up to the post-processing logic.

amorenoz replied:
> > Hmm... zero is still a valid smp id so I don't know if we should do that. Maybe having it `Optional`? Other than that, I think it's a good idea. I don't really like having a json object with just one field but I think it's better than having a "common" section that is not common.
> 
> Sounds good.
> 
> > Some other thought (trying to think of other pieces of information that might come handy in a "common" section): Do we need to flag an event as a metadata one? Events coming from eBPF side won't have metadata sections right? If a collector has to add data that is related to it, it can create another (normal) section.
> > So, maybe we need a `type=[bpf|meta]` field? or even a `meta=[true|false] # default false to avoid extra space in normal events which are more common`?
> 
> (Note that before writing this down I already made the common section to be mandatory only).
> 
> That would shift the metadata vs non-metadata to the event instead of the section level. My idea was really not to differentiate "metadata" vs other sections. However I do agree we _do_ make a distinction as of now (more on that later).
> 
> I don't think an event should be metadata only or metadata free. The metadata is there to provide additional information for post-processing. It can be global information (that can be extended / amended by future events) but it could be specific to a given event.
> 

> Taking a step back, the main points to define to get an agreement here are: what is an event and what is a metadata section.
> 
> For the metadata part, my view is this should be at the section level. The current distinction based on the section id might not be the best fit, neither is the "metadata" word we use in some places. IMO the only difference is the ability for `EventFmt` to produce something. Maybe we should only make the distinction here (it's already done in `PrintSingle/Series` for example): a "metadata section" would be a normal section not producing any direct output.
> 

I  don't think we should make the distinction at the `EventFmt` neither. I can very well see interest in output formats with the need to print metadata sections, specially with sections added by post-processing tools that enrich an event that came from the kernel.

How about we do as with `skb` sections? In commands such as `print` and `sort`, we allow the user to define what sections to print, the default being the ones from enabled collectors (they were explicitly enabled so we can assume some interest from the user). We expose an option, for now: `--show-extra-sections=[global]`.
In `EventFmt` we check if sections were not enabled, we skip it.

WDYT?
 
> For the event part, I think it has to bet loosely defined to get some flexibility. To me an event is a set of sections somehow related to each others. It might be to a packet event (aka. probe + timestamp) or to the initial configuration that is dumped by collectors (aka. timestamp only).
> 
> The difference in how those are consumed at post-processing time (is it global info? specific to an event?) is really only depending on the kind of section it is. That's up to the post-processing logic.

Apart from the comment above which I don't think changes the essence of it, I agree.

atenart replied:
> > Taking a step back, the main points to define to get an agreement here are: what is an event and what is a metadata section.
> > For the metadata part, my view is this should be at the section level. The current distinction based on the section id might not be the best fit, neither is the "metadata" word we use in some places. IMO the only difference is the ability for `EventFmt` to produce something. Maybe we should only make the distinction here (it's already done in `PrintSingle/Series` for example): a "metadata section" would be a normal section not producing any direct output.
> 
> I don't think we should make the distinction at the `EventFmt` neither. I can very well see interest in output formats with the need to print metadata sections, specially with sections added by post-processing tools that enrich an event that came from the kernel.

Right. Let me rephrase it: I don't think we should distinguish any metadata event from non-metadata ones. Everything is just an event. Some are linked to packets, some are not. But only a subset of sections is worth displaying *ever* in `collect`, `print` and `sort`. So we can actually make the distinction here (and not on the section being "metadata" or not). If a section is not meant to be displayed *ever*, let's not implement formatting logic for it.

> How about we do as with `skb` sections? In commands such as `print` and `sort`, we allow the user to define what sections to print, the default being the ones from enabled collectors (they were explicitly enabled so we can assume some interest from the user). We expose an option, for now: `--show-extra-sections=[global]`. In `EventFmt` we check if sections were not enabled, we skip it.
> 
> WDYT?

I don't think we have an example requiring this for now (introduced "md sections" are not providing direct value to the user in their own). But yes, I think it's a good way forward if we ever have some sections that might or might not be displayed, depending on what one wants. To be explicit, I think this aligns perfectly with the above proposal too, but would have no user right now.

amorenoz replied:
> > > Taking a step back, the main points to define to get an agreement here are: what is an event and what is a metadata section.
> > > For the metadata part, my view is this should be at the section level. The current distinction based on the section id might not be the best fit, neither is the "metadata" word we use in some places. IMO the only difference is the ability for `EventFmt` to produce something. Maybe we should only make the distinction here (it's already done in `PrintSingle/Series` for example): a "metadata section" would be a normal section not producing any direct output.
> > 
> > 
> > I don't think we should make the distinction at the `EventFmt` neither. I can very well see interest in output formats with the need to print metadata sections, specially with sections added by post-processing tools that enrich an event that came from the kernel.
> 
> Right. Let me rephrase it: I don't think we should distinguish any metadata event from non-metadata ones. Everything is just an event. Some are linked to packets, some are not. But only a subset of sections is worth displaying _ever_ in `collect`, `print` and `sort`. So we can actually make the distinction here (and not on the section being "metadata" or not). If a section is not meant to be displayed _ever_, let's not implement formatting logic for it.
> 

Sure but now all sections have formatting implemented, right? I am struggling to find a case where a section will not be printed in some way or another. Even config changes will need a way to be displayed, no?

> > How about we do as with `skb` sections? In commands such as `print` and `sort`, we allow the user to define what sections to print, the default being the ones from enabled collectors (they were explicitly enabled so we can assume some interest from the user). We expose an option, for now: `--show-extra-sections=[global]`. In `EventFmt` we check if sections were not enabled, we skip it.
> > WDYT?
> 
> I don't think we have an example requiring this for now (introduced "md sections" are not providing direct value to the user in their own). But yes, I think it's a good way forward if we ever have some sections that might or might not be displayed, depending on what one wants. To be explicit, I think this aligns perfectly with the above proposal too, but would have no user right now.

Fair enough

atenart replied:
> > > I don't think we should make the distinction at the `EventFmt` neither. I can very well see interest in output formats with the need to print metadata sections, specially with sections added by post-processing tools that enrich an event that came from the kernel.
> > 
> > 
> > Right. Let me rephrase it: I don't think we should distinguish any metadata event from non-metadata ones. Everything is just an event. Some are linked to packets, some are not. But only a subset of sections is worth displaying _ever_ in `collect`, `print` and `sort`. So we can actually make the distinction here (and not on the section being "metadata" or not). If a section is not meant to be displayed _ever_, let's not implement formatting logic for it.
> 
> Sure but now all sections have formatting implemented, right? I am struggling to find a case where a section will not be printed in some way or another. Even config changes will need a way to be displayed, no?

In the current form of this PR yes, but there's no reason really to implement `EventFmt` for the (currently named) `GlobalEventMd` section.

Another example I have in mind would be to dump the nft config in an event (to match at post-processing time the exact rule being hit) and here again I don't see a reason to display this in the events. The information might be used to augment what we display in some way, but the section itself has no reason to be displayed. I'm thinking this should actually apply to many sections of that kind.

Then of course we might see some sections where we'd like to print something; for those we can implement a way to select what to print.

atenart commented on retis/src/process/display.rs:
```diff
@@ -59,43 +62,38 @@ impl PrintSeries {
+        }
+
+        let mut res = Vec::new();
+        let mut delim = "+ ";
```
That would look quite good IMO, I'll make the change.

atenart commented on retis/src/process/display.rs:
```diff
@@ -5,7 +5,7 @@ use anyhow::Result;
 use crate::events::*;
 
-enum PrintSingleFormat {
+pub(crate) enum PrintSingleFormat {
```
Makes sense, I'll fix that.

atenart commented on retis/src/process/display.rs:
```diff
@@ -17,18 +17,8 @@ pub(crate) struct PrintSingle {
 }
 
 impl PrintSingle {
```
That makes perfect sense, will do.

atenart commented on retis/src/collect/collector.rs:
```diff
@@ -513,7 +526,12 @@ mod tests {
+            &mut self,
+            _: &CliConfig,
+            _: &mut ProbeBuilderManager,
+            _: &mut Event,
```
Collectors do not generate events, but event sections. Then can generate 0:n sections and put them into a single event. Also the initialization is done at a single point in time so there's no real need to have more than one event IMO, as all sections can be part of the same. Or do you have a use case in mind?

amorenoz replied:
I'm not sure I have a particular use case in mind. But there is something that does not quite fit.

Normal events (from eBPF) have different event sections. These sections are unrelated to each other (modular) but there is something that ties them together (into an event): they all refer to the same packet.

With metadata events we are adding 100% unrelated data into the same event which feels kinda weird, specially if the event itself has no semantics at all that ties the sections together (the "packet" for ebpf events).

IIUC, the idea is for modules to also generate md events during collection. When that happens, will we collect all event_sections into a single event? or will we have events with a single section? I guess the later, so this makes squeezing all event sections into a single event a bit weird.

Have you considered adding something that indicates that the first event is that, an "initial-configuration" event? That  would "tie" all sections together (they are each module's initial config) and would not seem so loose.

An alternative is to have each module generate a full event.

atenart replied:
> Normal events (from eBPF) have different event sections. These sections are unrelated to each other (modular) but there is something that ties them together (into an event): they all refer to the same packet.
> 
> With metadata events we are adding 100% unrelated data into the same event which feels kinda weird, specially if the event itself has no semantics at all that ties the sections together (the "packet" for ebpf events).
> 
> IIUC, the idea is for modules to also generate md events during collection. When that happens, will we collect all event_sections into a single event? or will we have events with a single section? I guess the later, so this makes squeezing all event sections into a single event a bit weird.

Discussion is a bit mangled between here and the other thread, please see my answer in there.

> Have you considered adding something that indicates that the first event is that, an "initial-configuration" event? That would "tie" all sections together (they are each module's initial config) and would not seem so loose.

The event being an "initial configuration" one comes from the fact that `CommonEventMd` is present. That is vary much like what's make a BPF event: the section about the probe is there. (In addition both have a `CommonEvent` section).

> An alternative is to have each module generate a full event.

I'm not 100% against that TBF.

amorenoz replied:
> > Normal events (from eBPF) have different event sections. These sections are unrelated to each other (modular) but there is something that ties them together (into an event): they all refer to the same packet.
> > With metadata events we are adding 100% unrelated data into the same event which feels kinda weird, specially if the event itself has no semantics at all that ties the sections together (the "packet" for ebpf events).
> > IIUC, the idea is for modules to also generate md events during collection. When that happens, will we collect all event_sections into a single event? or will we have events with a single section? I guess the later, so this makes squeezing all event sections into a single event a bit weird.
> 
> Discussion is a bit interleaved between here and the other thread, please see my answer in there.
> 

Sorry for the mess.

> > Have you considered adding something that indicates that the first event is that, an "initial-configuration" event? That would "tie" all sections together (they are each module's initial config) and would not seem so loose.
> 
> The event being an "initial configuration" one comes from the fact that `CommonEventMd` is present. That is vary much like what's make a BPF event: the section about the probe is there. (In addition both have a `CommonEvent` section).
> 

I agree that we should keep events loosely defined (see other thread).

So, let's see what kind of events we can have:

**A** - Events with probe section: What all the sections have in common is that they somehow describe the event that was originated by such probe. These event sections that come from the BPF side but also metadata events originated by, for instance, post-processing tools that provide extra data to the event.

**B** - Events without probe section: What these events have in common is the timestamp, (part of the `CommonEvent` section).

For events of type B, one could argue that, if some sections take some time to generate, the timestamp will be more accurate for some sections than others. So I see this "connection" between sections kind of loose (or at least more loose than for events with a probe section). We could add a section (the equivalent for the probe section for these events) that defines the "stage" this event was generated at, (e.g: init, end, ... ). However, this will make spontaneous events generated during collection (e.g: due to some config change) also kind of loose and only bound by the timestamp. So I guess I'll have to live with this as I don't have a better solution.

atenart replied:
> > Discussion is a bit interleaved between here and the other thread, please see my answer in there.
> 
> Sorry for the mess.

No worries, this is only because of the GH interface :)

> > > Have you considered adding something that indicates that the first event is that, an "initial-configuration" event? That would "tie" all sections together (they are each module's initial config) and would not seem so loose.
> > 
> > The event being an "initial configuration" one comes from the fact that `CommonEventMd` is present. That is vary much like what's make a BPF event: the section about the probe is there. (In addition both have a `CommonEvent` section).
> 
> I agree that we should keep events loosely defined (see other thread).
> 
> So, let's see what kind of events we can have:
> 
> **A** - Events with probe section: What all the sections have in common is that they somehow describe the event that was originated by such probe. These event sections that come from the BPF side but also metadata events originated by, for instance, post-processing tools that provide extra data to the event.
> 
> **B** - Events without probe section: What these events have in common is the timestamp, (part of the `CommonEvent` section).

And that's only for now. I would just summarize events as being a collection of sections sharing something in common. Be it a probe + ts or something else. There's no real concept of "metadata" in reality (I'm trying to stop using the word).

> For events of type B, one could argue that, if some sections take some time to generate, the timestamp will be more accurate for some sections than others. So I see this "connection" between sections kind of loose (or at least more loose than for events with a probe section). We could add a section (the equivalent for the probe section for these events) that defines the "stage" this event was generated at, (e.g: init, end, ... ). However, this will make spontaneous events generated during collection (e.g: due to some config change) also kind of loose and only bound by the timestamp. So I guess I'll have to live with this as I don't have a better solution.

I do agree that "config" sections only sharing a timestamp is the one case not perfectly fitting. We can still issue a single event per "logical data", aka. allowing modules to issue a full event and not a section only at init time. This would also map better to what could happen later: modules issuing an update event sections for some kind of config. A full event would probably be generated in such case.

amorenoz replied:
> I do agree that "config" sections only sharing a timestamp is the one case not perfectly fitting. We can still issue a single event per "logical data", aka. allowing modules to issue a full event and not a section only at init time. This would also map better to what could happen later: modules issuing an update event sections for some kind of config. A full event would probably be generated in such case.

I guess so, yes. That's why I originally suggested it. Either way is fine, specially since we don't have per-module events and I cannot see a practical example.

atenart commented on retis-events/src/display.rs:
```diff
@@ -1,12 +1,24 @@
 
+#[derive(Debug, Default, Clone, Copy)]
+pub struct DisplayFormat {
+    pub flavor: DisplayFormatFlavor,
```
I was wondering whether or not to remove `DisplayFormatFlavor` entirely. I still don't have a strong opinion on this. The reasoning was this maps well to the cli `--format` option and is a top level configuration of the `DisplayFormat`, other embedded flags are more fine grained. In case we have more top-level configuration in the future, having an helper returning a `bool` wouldn't be easy to consume too. Or to say it differently, `DisplayFormatFlavor` would apply to everything in an event while embedded flags/values in `DisplayFormat` ar only for some parts.

But as said, no strong opinion.

amorenoz replied:
I guess that distinction does exist (whether it applies to the whole event or not) but as soon as some other option that applies to the full event is added, if it's orthogonal to multi/single line it will feel strange. E.g: `color: bool`.

atenart replied:
You had a point, makes sense.

atenart commented on retis-events/src/time.rs:
```diff
@@ -0,0 +1,74 @@
+        DateTime::from_timestamp(val.sec(), val.nsec() as u32)
+            .expect("Could not convert TimeSpec to DateTime")
+    }
+}
```
Makes sense, I'll add some.

atenart commented on retis-events/src/common.rs:
```diff
@@ -2,6 +2,17 @@ use std::fmt;
 
+#[event_section("md_common")]
+pub struct CommonEventMd {
+    pub retis_version: String,
```
Yep. I'm also wondering if this should not be named "global" instead of "common" as it's a section not common to all metadata events but just global information available for everyone. WDYT?

amorenoz replied:
+1 to not use "common" again :-)

atenart commented on retis-events/src/events.rs:
```diff
@@ -267,13 +271,14 @@ impl SectionId {
     /// Is the section a metadata one?
     pub fn is_metadata(&self) -> bool {
-        false
+        matches!(self, SectionId::MdCommon)
```
Makes sense, we already have "special sections" at the beginning, then "collector sections" so adding one "md sections" looks good. The only thing is extra maintenance when adding a collector section, but that's not a big issue.

atenart commented:
> Also, we might need a section in the docs for the new event section?

Yep. We'd need a new page about the event sections too as only the collector sections are described as of now. I'll add it.

atenart commented:
Pushed with most of the comments (multiline format, docs, optional smp_id, etc) taken into account. It doesn't contain modifications related to the nature of md events, as discussion is ongoing.

amorenoz commented on retis-events/src/common.rs:
```diff
@@ -19,13 +37,27 @@ pub struct CommonEvent {
     /// SMP processor id.
-    pub smp_id: u32,
+    pub smp_id: Option<u32>,
     pub task: Option<TaskEvent>,
```
Unrelated but "task" does not have a comment.

atenart commented:
To sum up the current state of this PR, two modifications are needed:
1. Stop putting event sections generated by the collectors in the same event (start time or later).
1. Find a way not to display all event sections by default, or at least for now.

Thinking more about the two above points, some things are missing in this PR to have a clean implementation.

For point 1, the current state giving a single `mut` event to all collectors was not perfect (they shouldn't be allowed to modify the other sections) but when moving towards a list of events (so each collector can have its own) things get even worse. For this to be correctly supported I think we need to introduce a "Rust event factory", that we could consume the same way we consume the raw factory. This factory should be thread safe and collectors should be able to keep a reference (for later updates, if needed). That should make things clean.

(We don't strictly need the above as only the core is generating an event section right now, but not introducing this would feel weird and not allow any future user).

For point 2 things are not straightforward. My initial thought was to let some section not implement `EventFmt` but then this makes sections not to follow a common implementation and that is quite an issue. We could do things at the `EventDisplay` level but that has the same kinds of disadvantages, including not being compatible with #411. My impression so far is @amorenoz 's proposal of controlling which sections should be printed or not is probably best. There are some gotchas (we probably don't want to make all sections optional) but overall this should be doable. This should be easier to do with #411 though... which depends on this PR. For the time being, and while waiting for an actual use case we can extend our logic of sorting `SectionId`. It currently has two states: special handling and automatic handling; and could have an extra state of sections being excluded from formatting by default.

atenart commented:
> To sum up the current state of this PR, two modifications are needed:
> 
>     1. Stop putting event sections generated by the collectors in the same event (start time or later).
> 
>     2. Find a way not to display all event sections by default, or at least for now.
> 
> Thinking more about the two above points, some things are missing in this PR to have a clean implementation.
> 
> For point 1, the current state giving a single `mut` event to all collectors was not perfect (they shouldn't be allowed to modify the other sections) but when moving towards a list of events (so each collector can have its own) things get even worse. For this to be correctly supported I think we need to introduce a "Rust event factory", that we could consume the same way we consume the raw factory. This factory should be thread safe and collectors should be able to keep a reference (for later updates, if needed). That should make things clean.
> 
> (We don't strictly need the above as only the core is generating an event section right now, but not introducing this would feel weird and not allow any future user).

I introduced a `RetisEventsFactory` as an API to let parts of Retis (core, collectors, etc) add their own events. The way it is defined allows for its use across different threads. One note about this is there could be some events being reordered, as the raw factory can wait up to a second for incoming events. It's not perfect but IMO this is a broader topic we'll also have to handle if we split raw event processing across different threads at some points.

(For reference I don't think this can be solved at the event dequeuing part -`next_event`- as there is always a possibility to see races. With that either we introduce a small buffer and reorder events on the fly before feeding the printers, or we sort events by default at post-processing time -sort, not group like in the `sort` command-).

> For point 2 things are not straightforward. My initial thought was to let some section not implement `EventFmt` but then this makes sections not to follow a common implementation and that is quite an issue. We could do things at the `EventDisplay` level but that has the same kinds of disadvantages, including not being compatible with #411. My impression so far is @amorenoz 's proposal of controlling which sections should be printed or not is probably best. There are some gotchas (we probably don't want to make all sections optional) but overall this should be doable. This should be easier to do with #411 though... which depends on this PR. For the time being, and while waiting for an actual use case we can extend our logic of sorting `SectionId`. It currently has two states: special handling and automatic handling; and could have an extra state of sections being excluded from formatting by default.

I removed entirely conditionals to print or not print events based on their sections. We currently only have a single non-raw section and as discussed with @amorenoz we actually might want to display it. In any way, it's fine for now. I think the way forward for this is to work on #411 and then explicit the list of sections we want to see. This will need some thinking to have a clean implementation.

In addition to the above I renamed the "global" event section into "startup" and added comments on the task event section.

My only concern would be about the event reordering, although nothing would be impacted here. I believe this is a separate topic that needs some discussion and I think this series is ready; but this is not a  strong opinion, if you consider this should be solved here.

amorenoz commented on retis/src/process/cli/print.rs:
```diff
@@ -26,6 +26,9 @@ pub(crate) struct Print {
     #[clap(value_enum, default_value_t=CliDisplayFormat::MultiLine)]
     pub(super) format: CliDisplayFormat,
+    #[arg(long, help = "Time format used when printing an event.")]
+    #[clap(value_enum, default_value_t=CliTimeFormat::MonotonicTimestamp)]
```
nit: I don't see users typing such a long cli option "--time-format=monotonic-offset" (plus, the word "monotonic" might not mean a lot to many). From a purely user experience pov, how about a flag `--utc` to achieve the same?

atenart replied:
The initial reasoning was to allow other variant of the time format (as a side note I would not expect anyone to set `--time-format=monotonic-offset` as it's the default value). But I'm also not sure what would be those other time formats :)

I think using `--utc` is fine and better for UX.

atenart replied:
Do you also suggest to remove `TimeFormat` in `DisplayFormat` and use a simple `utc: bool`?

amorenoz commented on retis/src/collect/cli.rs:
```diff
@@ -68,6 +68,9 @@ defaults to \"retis.data\"."
     pub(super) format: CliDisplayFormat,
+    #[arg(long, help = "Time format used when printing an event.")]
+    #[clap(value_enum, default_value_t=CliTimeFormat::MonotonicTimestamp)]
+    pub(super) time_format: CliTimeFormat,
```
ditto

amorenoz commented on retis/src/benchmark/events_output.rs:
nit in the commit message of patch 20. s/This is/This

amorenoz reviewed (commented):
I have some small comments but overall the PR looks awesome. I have played with it and it's great.

amorenoz commented on retis/src/core/events/factory.rs:
```diff
@@ -0,0 +1,52 @@
+impl RetisEventsFactory {
+    /// Retrieve the next event, if any. Returns EventResult::Retry if no event
+    /// is currently available.
+    pub(crate) fn next_event(&self, _: Option<Duration>) -> Result<EventResult> {
```
BTW, I can do with this struct the same I do with `FileEventsFactory` in #426 , meaning we can fix it in a different patch if you like.

atenart replied:
That makes perfect sense, let's not introduce yet another example of this. I'll remove it from that PR.

atenart commented:
@amorenoz thanks for the reviews and the good suggestions! I took all into account and had to rebase onto main to resolve a conflict.

atenart commented:
As usual, some weeks rawhide does not issue a libvirt image. This is why all the runtime tests are failing.

amorenoz reviewed (approved):
Looks good to me, thanks.
